File: app/models/opportunity.py
from sqlalchemy import Column, Integer, String, Float, Boolean, DateTime, Text, UniqueConstraint
from sqlalchemy.dialects.sqlite import JSON
from datetime import datetime
from app.database import Base


class Opportunity(Base):
    __tablename__ = "opportunities"
    __table_args__ = (
        UniqueConstraint("source", "external_id", name="uq_opportunity_source_external"),
    )

    id = Column(Integer, primary_key=True, index=True)

    # Basic info
    title = Column(String, nullable=False)
    description = Column(Text)
    required_skills = Column(JSON, default=list)  # Normalized skills list
    preferred_skills = Column(JSON, default=list)  # Nice-to-have skills
    category = Column(String)

    # Company info
    company_name = Column(String)
    company = Column(String)  # alias for external sources; keep both for compatibility
    company_logo_url = Column(String)
    company_website = Column(String)
    company_size = Column(String)  # "startup", "small", "medium", "large", "enterprise"

    # Location
    location = Column(String)
    city = Column(String)
    state = Column(String)
    country = Column(String)
    is_remote = Column(Boolean, default=False)

    # Compensation
    salary_min = Column(Float)
    salary_max = Column(Float)
    salary_currency = Column(String, default="USD")
    salary_period = Column(String, default="yearly")  # "yearly", "monthly", "hourly"
    is_salary_visible = Column(Boolean, default=True)

    # Classification
    job_type = Column(String)  # "fulltime", "parttime", "internship", "contract"
    opportunity_type = Column(String, default="job")  # "job", "internship", "scholarship", "grant"
    work_arrangement = Column(String, default="onsite")  # "remote", "hybrid", "onsite", "contract"
    experience_level = Column(String, default="entry")  # "entry", "mid", "senior", "executive"
    education_requirement = Column(String, default="none")  # "none", "high_school", "associate", "bachelors", "masters", "phd"

    # Application details
    url = Column(String)  # external apply URL
    application_url = Column(String)
    application_email = Column(String)
    application_deadline = Column(DateTime)

    # Metadata
    posted_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    created_at = Column(DateTime, default=datetime.utcnow)
    expires_at = Column(DateTime)
    is_active = Column(Boolean, default=True)
    is_featured = Column(Boolean, default=False)

    # Benefits
    benefits = Column(JSON, default=list)  # ["health_insurance", "401k", "remote_work", etc.]

    # For scholarships/grants specifically
    award_amount = Column(Float)
    award_currency = Column(String)
    eligibility_criteria = Column(JSON, default=list)

    # External sourcing metadata
    source = Column(String, default="internal")
    external_id = Column(String)
    external_url = Column(String)
    refreshed_at = Column(DateTime)
    is_stale = Column(Boolean, default=False)

    def mark_refreshed(self):
        self.refreshed_at = datetime.utcnow()
        self.is_stale = False

    def mark_stale(self):
        self.is_stale = True

File: tender-frontend/src/types/opportunity.ts
export type OpportunityType = 'job' | 'internship' | 'scholarship' | 'grant';
export type WorkArrangement = 'remote' | 'hybrid' | 'onsite';
export type ExperienceLevel = 'entry' | 'mid' | 'senior' | 'executive';

export interface Opportunity {
  id: number;
  title: string;
  description: string;
  required_skills: string[];
  category?: string;

  // Company info
  company_name?: string;
  company_logo_url?: string;
  company_website?: string;
  company_size?: string;

  // Location
  city?: string;
  state?: string;
  country?: string;
  is_remote: boolean;

  // Compensation
  salary_min?: number;
  salary_max?: number;
  salary_currency: string;
  salary_period: string;
  is_salary_visible: boolean;

  // Classification
  opportunity_type: OpportunityType;
  work_arrangement: WorkArrangement;
  experience_level: ExperienceLevel;
  education_requirement?: string;

  // Application
  application_url?: string;
  application_email?: string;
  application_deadline?: string;

  // Metadata
  posted_at?: string;
  updated_at?: string;
  expires_at?: string;
  is_active: boolean;
  is_featured: boolean;

  // Extra
  preferred_skills: string[];
  benefits: string[];
  award_amount?: number;

  // Enhanced details for comprehensive job description
  desired_qualifications?: string[];
  work_authorization?: string[];  // e.g., ['US Citizens', 'Green Card Holders', 'OPT/CPT Eligible']
  country_restrictions?: string[];  // Countries where candidates must be authorized to work
  open_to_international?: boolean;
  availability_start?: string;  // ISO date string
  availability_end?: string;  // ISO date string
  graduation_year_min?: number;  // e.g., 2025
  graduation_year_max?: number;  // e.g., 2027
  employer_requirements?: string;  // Additional requirements from employer
  responsibilities?: string[];  // Job responsibilities
}

export interface OpportunityMatch extends Opportunity {
  match_score: number;
  matched_skills: string[];
}
